application/src/main/kotlin/net/corda/v5/application/crypto/MerkleTreeFactory.kt

```diff
@@ -0,0 +1,43 @@
+package net.corda.v5.application.crypto
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.base.annotations.DoNotImplement
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.crypto.merkle.MerkleTree
+import net.corda.v5.crypto.merkle.MerkleTreeHashDigest
+
+/**
+ * [MerkleTreeFactory] creates [MerkleTree]s and [MerkleTreeHashDigest]s.
+ */
+@DoNotImplement
+interface MerkleTreeFactory {
+    /**
+     * Creates a [MerkleTree]
+     *
+     * @param leaves The leaves of the tree.
+     * @param digestProvider Merkle Tree Hash digest provider used to construct the tree's node and leaf hashes.
+     *
+     * @return A new [MerkleTree] instance.
+     */
+    @Suspendable
+    fun createTree(
+        leaves: List<ByteArray>,
+        digestProvider: MerkleTreeHashDigest
+    ) : MerkleTree
+
+    /**
+     * Creates a [MerkleTreeHashDigest].
+     *
+     * @param merkleTreeHashDigestProviderName name of the Hash Digest Provider class
+     * @param digestAlgorithmName name of the base Hash algorithm
+     * @param options Hash Digest provider specific options
+     *
+     * @return A new [MerkleTreeHashDigest] instance.
+     */
+    @Suspendable
+    fun createHashDigestProvider(
```

Comment:
@omfR3 can you rename it to `createHashDigest`?

Reply:
Yes. I still don't quite understand whether that interface is supposed to represent a hash digest or a hash digest provider, but I'll make that change.